File: leet/arrays_and_hashing/two_sum.py
class Solution:
    # O()
    # assume: always 1 solution
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        d = dict()

        for i, num in enumerate(nums):
            need = target - num
            if d.get(need) != None:
                return [d[need], i]
            else:
                d[num] = i

s = Solution()

nums1 = [2,7,11,15]
t1 = 9

s.twoSum(nums1, t1)

File: leet/arrays_and_hashing/two_sum.cpp
// https://leetcode.com/problems/two-sum/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        int n = nums.size();
        unordered_map<int, int> seen;

        for (int i = 0; i < n; i++) {
            int x = nums[i];
            int need = target-x;
            if (seen.count(need)) {
                return {i, seen[need]};
            }
            seen[x] = i;
        }

        return {-1, -1}; // shouldnt happen by problem statement
    }
};